[AIRFLOW-6535] add exception to fail without retry

jstern opened the pull request:
I would like to be able to configure certain tasks to retry, but to also have a way to bypass retry if I can detect a condition that is unlikely to be change for the better.

For example, imagine I have a DAG with a large number of tasks that fetch data from an API ... sometimes the API returns a 200 and everything is fine, sometimes the API returns a 500 and I know this means that API is failing under load and I want to retry later, and sometimes the API returns a 400 indicating that I've configured my requests in a way that will never succeed. If the API returns a 400 for some/all of my tasks, then I have to wait for all of them to get through all their retries before the run fails.

My proposal is to add another subclass of AirflowException called AirflowFailException, and to update the exception handling in TaskInstance.run_raw_task such that when this exception is seen, the resulting behavior is that same as if we entered handle_failure and is_eligible_for_returned false (except with adjusted logging so it does not look like we simply hit our retry limit). 

---
Issue link: [AIRFLOW-6535](https://issues.apache.org/jira/browse/AIRFLOW-6535)

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

Commit:
first pass at exception to fail without retry

potiuk commented on airflow/models/taskinstance.py:
```diff
@@ -1120,10 +1125,33 @@ def handle_failure(self, error, test_mode=None, context=None, session=None):
-                self.log.info('Marking task as UP_FOR_RETRY')
-                if task.email_on_retry and task.email:
-                    self.email_alert(error)
+                if force_fail is not True:
```
```suggestion
                if not force_fail:
```

or even better reverse the if and use `if force_fail:`

potiuk reviewed (commented):
I like it, just one small comment. It looks like a good way. However it will also need some description in the documentation and some tests.

jstern committed:
use preferred boolean check idiom

Co-Authored-By: Jarek Potiuk <jarek@potiuk.com>

Commit:
add test coverage for AirflowFailException

Commit:
add docs for some exception usage patterns

Commit:
autoformatting

Commit:
remove extraneous newline, poke travis build

houqp commented on airflow/models/taskinstance.py:
```diff
@@ -1120,10 +1125,33 @@ def handle_failure(self, error, test_mode=None, context=None, session=None):
-                self.log.info('Marking task as UP_FOR_RETRY')
-                if task.email_on_retry and task.email:
-                    self.email_alert(error)
+                if not force_fail:
```
perhaps we can make this if else combo less nested to reduce the complexity:

```python
if not force_fail and self.is_eligible_to_retry():
    self.state = State.UP_FOR_RETRY
    ...
else:
    if force_fail:
       info_msg = "Immediate failure requested; skipping retries and marking task as FAILED."
    elif task.retries:
       info_msg = "All retries failed; marking task as FAILED."
    else:
       info_msg = "Marking task as FAILED."
    self.log.info("%s.\ndag_id=%s, task_id=%s, execution_date=%s, start_date=%s, end_date=%s", info_msg, ...)
    if task.email_on_failure and task.email:
        self.email_alert(error)
```

I think it's reasonable to not check `task.retries` when force_fail is set and always show `Immediate failure requested; skipping retries and marking task as FAILED.`. If you still want to keep it, we can add that in `if force_fail` block.

jstern replied:
yeah, the messiness of the nesting was weighing on me a bit when i made [this comment](https://github.com/apache/airflow/pull/7133#issuecomment-573155093) above, but what I was thinking of then wouldn't have worked because we still want all the callbacks to fire. I'll take a look at implementing your suggestion

Commit:
clean up TaskInstance.handle_failure

Try to reduce nesting and repetition of logic for different conditions.
Also try to tighten up the scope of the exception handling ... it looks
like the large block that catches an Exception and logs it as a failure
to send an email may have been swallowing some TypeErrors coming out
of trying to compose a log info message and calling strftime on
start_date and end_date when they're set to None; this is why I've added
lines in the test to set those values on the TaskInstance objects.

Commit:
let sphinx generate docs for exceptions module

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1124,64 +1129,56 @@ def handle_failure(self, error, test_mode=None, context=None, session=None):
             context['exception'] = error
 
-        # Let's go deeper
-        try:
```
What happened to this try/except?

jstern replied:
the try/excepts have been tightened in scope to only be around the conditions they are logging and moving on from. see my response to your question about the start/end dates in the tests for more color on that

jstern replied:
To give a more complete answer, here is the try/except in master. I've elided some arguments to log.info and added comments to the two lines where an exception being raised **would** result in an accurately logged error message. Exceptions raised on other lines can cause us to log a message about failing to send an email when that's not what actually happened.

```python
        # Let's go deeper
        try:
            # Since this function is called only when the TaskInstance state is running,
            # try_number contains the current try_number (not the next). We
            # only mark task instance as FAILED if the next task instance
            # try_number exceeds the max_tries.
            if self.is_eligible_to_retry():
                self.state = State.UP_FOR_RETRY
                self.log.info('Marking task as UP_FOR_RETRY')
                if task.email_on_retry and task.email:
                    self.email_alert(error)  # RELEVANT TO EXCEPT BLOCK
            else:
                self.state = State.FAILED
                if task.retries:
                    self.log.info(...)
                else:
                    self.log.info(...)
                if task.email_on_failure and task.email:
                    self.email_alert(error)  # RELEVANT TO EXCEPT BLOCK
        except Exception as e2:
            self.log.error('Failed to send email to: %s', task.email)
            self.log.exception(e2)
```

And here is the more tightly-scoped version in this PR:

```python
    if email_for_state and task.email:
            try:
                self.email_alert(error)
            except Exception as e2:
                self.log.error('Failed to send email to: %s', task.email)
                self.log.exception(e2)
```

While https://github.com/apache/airflow/pull/7257 addresses the only non-email failure I've personally seen happen in practice, this PR does put some lines of code that were previously inside of a try/except on the outside (and I'm sure what I've personally seen happen is a fraction of what actually might). Please let me know if this increased chance of other exceptions bubbling up uncaught is unacceptable and I'll look to restructure while trying to preserve accurate logging....

ashb replied:
Does is_eligable_for_retry do anything more than check instance attributes? If not then this change is probably okay

jstern replied:
Here's the entirety of it:

```python
    def is_eligible_to_retry(self):
        """Is task instance is eligible for retry"""
        return self.task.retries and self.try_number <= self.max_tries
```

ashb replied:
Thanks (I was reviewing from my phone, so wasn't easy for me to check)

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1103,7 +1108,7 @@ def _handle_reschedule(self, actual_start_date, reschedule_exception, test_mode=
 
     @provide_session
-    def handle_failure(self, error, test_mode=None, context=None, session=None):
+    def handle_failure(self, error, test_mode=None, context=None, session=None, force_fail=False):
```
A style point: please keep session as the final argument.

jstern replied:
10-4, will change!

ashb commented on docs/conf.py:
```diff
@@ -207,9 +207,12 @@
 
 # Generate top-level
+
+allowed_top_level = ("exceptions.py",)
```
What is this all about?

jstern replied:
ah, was wondering if this would come up. @potiuk suggested that I should add documentation to the concepts section, and it felt incomplete to me to add that without linking to the exceptions api docs. all modules at the top level (airflow/*.py) are excluded from the documentation and my link to the exceptions api doc could not be rendered and was breaking the docs build. adding this ... exception :) ... fixed that

i'm happy to take any direction whether that's leaving this as-is, removing the links to the exception api docs, or something in-between. just let me know what you'd prefer

jstern replied:
i can also just add a comment explaining why this exists

ashb replied:
Ah, the API is handled differently (we render it with an autodoc extension) - can you get away with just linking to

```
   :py.module:`airflow.exceptions`
```

where you want to and then this change might not be needed?

ashb replied:
Ah, the API is handled differently (we render it with an autodoc extension) - can you get away with just linking to

```
   :py.module:`airflow.exceptions`
```

where you want to and then this change might not be needed?

jstern replied:
This works ... 

```
:py:mod:`airflow.exceptions`
```

... but only if I keep my changes in conf.py 😦 Otherwise it "works" but does not create a link. I'm pretty sure the code I'm touching in conf.py is actively excluding that module from the autodocs.

ashb replied:
@mik-laj @potiuk Any thoughts on this specific change?

ashb replied:
Looking at it a bit more I think you are right though @jstern

jstern replied:
@ashb @potiuk @mik-laj - any thoughts about whether it's ok to keep this change to include the exceptions module in the docs?

---

On a related note, after updating this pr from master, I'm getting a new docs build failure:

```
/opt/airflow/docs/_api/airflow/providers/amazon/aws/hooks/datasync/index.rst:66: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
/opt/airflow/docs/_api/airflow/providers/amazon/aws/operators/datasync/index.rst:167: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
/opt/airflow/docs/_api/airflow/providers/amazon/aws/operators/datasync/index.rst:167: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
/opt/airflow/docs/_api/airflow/providers/amazon/aws/operators/datasync/index.rst:167: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
/opt/airflow/docs/_api/airflow/providers/amazon/aws/operators/datasync/index.rst:167: WARNING: more than one target found for cross-reference 'AirflowException': airflow.exceptions.AirflowException, airflow.providers.yandex.hooks.AirflowException
```

I can make this go away by removing my change that allows `airflow.exceptions` api docs to be built, but this does not seem like a satisfactory course of action, because the `airflow.providers.yandex.hooks.AirflowException` will remain in the docs and it shouldn't. It's there because somehow these lines in `airflow/providers/yandex/hooks/__init__.py`:

```
from airflow.exceptions import AirflowException
from airflow.hooks.base_hook import BaseHook
```

are causing autoapi to include `AirflowException` and `BaseHook` in the docs for `airflow.providers.yandex.hooks` as if those classes were defined in that module.

<img width="719" alt="Screen Shot 2020-02-27 at 4 25 31 PM" src="https://user-images.githubusercontent.com/617540/75493344-46b1d680-597f-11ea-9e2c-2b8d30d3c302.png">

In other words, autoapi is generating this in `docs/_api/airflow/providers/yandex/hooks/index.rst`:

```
Package Contents
----------------

.. py:exception:: AirflowException

   Bases: :class:`Exception`

   Base class for all Airflow's errors.
   Each custom exception should be derived from this class

   .. attribute:: status_code
      :annotation: = 500

.. py:class:: BaseHook
```

I don't know if this is a bug in autoapi or in how the airflow project configures it, and I don't know why the warnings are only about AirflowException and not BaseHook...

ashb commented on tests/models/test_taskinstance.py:
```diff
@@ -1475,6 +1476,8 @@ def test_handle_failure(self):
         ti1 = TI(task=task1, execution_date=start_date)
         ti1.state = State.FAILED
+        ti1.start_date = start_date
+        ti1.end_date = end_date
```
Why was this change needed? The tests should pass without this change, no?

jstern replied:
the original structure of the method under test had a wide scope for the try/except block that was intended to log when we're unable to send an email and move on. i think TypeError's due to these attributes being none were being caught by that block and allowing the test to complete without error

jstern replied:
for more context, here is the failure that cropped up during the execution of existing test code after I narrowed the scope of the try/except to only be around sending the email. my mistake, it was an AttributeError not a TypeError :)

```
        self.log.info(
            '%s dag_id=%s, task_id=%s, execution_date=%s, start_date=%s, end_date=%s',
            log_message,
            self.dag_id,
            self.task_id,
            self.execution_date.strftime('%Y%m%dT%H%M%S'),
>           self.start_date.strftime('%Y%m%dT%H%M%S'),
            self.end_date.strftime('%Y%m%dT%H%M%S')
        )
E       AttributeError: 'NoneType' object has no attribute 'strftime'
```

jstern replied:
Aha! https://github.com/apache/airflow/pull/7257

ashb replied:
With that PR is this change now not needed? (Does it make sense to keep it anyway?)

jstern replied:
Actually, I needed to do something to make sure my boiling down of the logging.info call into one spot also incorporated that safety check. I've done that now and removed the start_date/end_date assignments from the test again

ashb reviewed (commented):
Nice idea.

The one thing I don't think the tests cover is that actually raising this new exception causes the task to not be retried -- you've tested the change to handle_failure, but that is essentially the internal details, rather than the "interface" of the new feature.

Commit:
keep session kwarg last in handle_failure

Commit:
explain allowed_top_level

Commit:
add black-box tests for retry/fail immediately cases

Commit:
don't lose safety measures in logging date attrs

Commit:
fix flake8 too few blank lines

Commit:
grammar nitpick

Commit:
Merge branch 'master' into airflow-fail-exception

ashb commented on docs/concepts.rst:
```diff
@@ -1216,6 +1216,49 @@ template string:
+This example illustrates some possibilities
+
+.. code:: python
+
```
can you add the import here? `from airflow.exceptions import ...`

jstern replied:
done :)

ashb reviewed (changes requested):
one tiny doc change then I'm happy with this.

sorry you had to chase so many times!

Commit:
add import to AirflowFailException example

Commit:
Merge branch 'airflow-fail-exception' of github.com:jstern/airflow into airflow-fail-exception
